### PR TITLE
Add OnBuild support for Dockerfiles

### PR DESCRIPTION
**Pull request by rhatdan**
Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Replaces https://github.com/projectatomic/buildah/pull/686

**Comment by rhatdan** [comment]
@TomSweeneyRedHat @nalind PTAL

**Comment by rhatdan** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
Fixes https://github.com/projectatomic/buildah/issues/658

**Comment by TomSweeneyRedHat** [comment]
@rhatdan I'll play more with this today, but it looks like you've the same issue I had.  But I think the config additions that you put in gave me a thought.

Here's the last bit of test results with your changes:

```
]# buildah bud -f ~/Dockerfile-2 -t result-image .
STEP 1: FROM onbuild-image
STEP 2: RUN touch /baz
STEP 3: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]localhost/result-image:latest
Getting image source signatures
Skipping fetch of repeat blob sha256:891e1e4ef82ad02a4ea1f030831f942d722c7694c4db64ca3239c8163b811c58
Skipping fetch of repeat blob sha256:4487d6d64f2494619f176ed275ad8fee4b737ef1017e1f43873a5cf1027e4052
Copying blob sha256:d3a7b3612d6d70d927c9664bb56e9d7881226e9d579309172ea772f1efedefca
 98 B / 98 B [==============================================================] 0s
Copying config sha256:18d727f1affda9561f1471e5261e08a9212ea8eadf34d9f2fd3955cb084f3816
 1.75 KiB / 1.75 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
18d727f1affda9561f1471e5261e08a9212ea8eadf34d9f2fd3955cb084f3816
[root@localhost buildah]# buildah from result-image
result-image-working-container
[root@localhost buildah]# buildah run result-image-working-container ls /bar
ls: cannot access '/bar': No such file or directory
```

**Comment by rhatdan** [comment]
It touched /baz not /bar.

**Comment by TomSweeneyRedHat** [comment]
@rhatdan, yep, the first Dockerfile touched 'foo', the second touched 'baz'.  However the OnBuild command in the first Dockerfile touched "bar' and that's what's missing in the resulting container.  I did the bar/baz dance a few times myself while testing.

```
# cat ~/Dockerfile
FROM fedora:28
RUN touch /foo
ONBUILD RUN touch /bar

# cat ~/Dockerfile-2
FROM onbuild-image
RUN touch /baz

# buildah run result-image-working-container ls /baz
/baz

# buildah run result-image-working-container ls /foo
/foo

# buildah run result-image-working-container ls /bar
ls: cannot access '/bar': No such file or directory

```

**Comment by TomSweeneyRedHat** [comment]
I played some more today.  I adjusted cmd/buildah/config.go to have it accept OnBuild in my sandbox.  I did it really cheesey, just doing this at the end of updateConfig():
```
        //TOM: delete next 2 lines
        tomString := []string {"RUN touch bar"}
        builder.SetOnBuild(tomString)
```
When I ran `buildah config` and then did `buildah inspect` of the newly created container, I saw "RUN touch bar" value in both OnBuild fields in the json.  I then did `buildah commit` and when I inspected the image both OnBuild fields were empty.  I think something might be going on in the commit code that's now quite right.  That's my next exploration.

This next part might be a red herring.  In the NewImageSource() function in the image.go file, I verified that we're dealing with a v1.MediaTypeImageManifest which stores aside the values from the oconfig.  The oci config does not have an OnBuild variable associated with it that I can find.  Should it?

Looking at the other possible manifest type there, docker.V25MediaTypeManifest, there is an OnBuild variable in there which I can add too, but that whole manifest is ignored in this scenario.

More digging tomorrow, thought I throw this out in case it uncovered anything.

**Comment by rhatdan** [comment]
Well OnBuild only works with doing --format docker.  Since OCI Does not understand OnBuild.  We are currently clearing the images OnBuild fields before adding new ones, which might be why it is dropping your config.  

BTW I will bin in Westford on Tuesday.  Maybe we can finish this up then.

**Comment by rhatdan** [comment]
@TomSweeneyRedHat This is now passing tests including the one for OnBuild and now you can add configuration via the buildah config --onbuild command.
PTAL

File: config.go [hunk 1]
```diff
@@ -331,6 +331,24 @@ func (b *Builder) SetUser(spec string) {
 	b.Docker.Config.User = spec
 }
 
+// OnBuild returns the OnBuild value from the container.
+func (b *Builder) OnBuild() []string {
+	return copyStringSlice(b.Docker.Config.OnBuild)
+}
+
+// ClearOnBuild removes all values from the OnBuild structure
```
Comment (TomSweeneyRedHat) [inline config.go]: Is this strictly necessary?  I wasn't running into an issue with left over cruft in the OnBuild.

File: docs/buildah-config.md [hunk 2]
```diff
@@ -82,6 +82,12 @@ Note: this setting is not present in the OCIv1 image format, so it is discarded
 Add an image *label* (e.g. label=*value*) to the image configuration of any
 images which will be built using the specified container. Can be used multiple times.
 
+**--onbuild** *onbuild command*
+
+Add an ONBUILD command to the image.  ONBUILD commands are automatically run
+when images are built based on the image you are creating.  ONBUILD images are
+only supported on `docker` formatted images.
+
```
Comment (TomSweeneyRedHat) [inline docs/buildah-config.md]: Might be nice to have an example for OnBuild.  Would it be good to add a note here to show how to determine what type of container it is?   A doc somewhere to show the difference between OCI and Docker formatted images and the commands applicable to each would be good.  Trello card for that?

Comment (TomSweeneyRedHat) [reply docs/buildah-config.md]: Nope, I missed adding the variable to the exec too.  Back to testing.

**Comment by TomSweeneyRedHat** [comment]
Testing looks good.  I want to play just a bit more:

```
# buildah bud -f ~/Dockerfile --format=docker -t onbuild-image .
STEP 1: FROM fedora:28
STEP 2: RUN touch /foo
STEP 3: ONBUILD RUN touch /bar
STEP 4: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]localhost/onbuild-image:latest
Getting image source signatures
Skipping fetch of repeat blob sha256:891e1e4ef82ad02a4ea1f030831f942d722c7694c4db64ca3239c8163b811c58
Copying blob sha256:dcce965cf0594c67e904f2837227e12f8276cc899231688a58f2451a01fb9da6
 141 B / 141 B [============================================================] 0s
Copying config sha256:009376b6ab528c2d4c8408ff2ee97aaead7c23d4e5bd1fa470fed7d4b601d388
 2.40 KiB / 2.40 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures

# buildah bud -f ~/Dockerfile-2 --format=docker -t result-image .
STEP 1: FROM onbuild-image
STEP 2: RUN touch /bar
STEP 3: RUN touch /baz
STEP 4: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]localhost/result-image:latest
Getting image source signatures
Skipping fetch of repeat blob sha256:891e1e4ef82ad02a4ea1f030831f942d722c7694c4db64ca3239c8163b811c58
Skipping fetch of repeat blob sha256:7abab12163e18b8b2918d3512fdeb52db86560d22eb146338dc3ef78ef46daf8
Copying blob sha256:5a41e19ec68e0213b95659d69ed94eff9fd30e718e6a1a7e49e472e66167196a
 112 B / 112 B [============================================================] 0s
Copying config sha256:bab16466828eda3ea8e28a80cf1f74126a818da6c9d9235b71319ea882036799
 2.60 KiB / 2.60 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
bab16466828eda3ea8e28a80cf1f74126a818da6c9d9235b71319ea882036799

# ctr=$(buildah from result-image)

# buildah run $ctr ls /bar
/bar
```

File: config.go (hunk 1 above)
Comment (TomSweeneyRedHat) [reply config.go]: FWIW, I tested without this function being called and it worked fine.  I don't know if you'd other scenarios in mind that this would handle.

File: imagebuildah/build.go [hunk 3]
```diff
@@ -758,6 +767,7 @@ func (b *Executor) Commit(ctx context.Context, ib *imagebuilder.Builder) (err er
 		PreferredManifestType: b.outputFormat,
 		IIDFile:               b.iidfile,
 		Squash:                b.squash,
+		OnBuild:               b.onbuild,
```
Comment (TomSweeneyRedHat) [inline imagebuildah/build.go]: I don't think this in necessary.  I commented it out and stuff worked fine.

File: imagebuildah/build.go [hunk 4]
```diff
@@ -183,6 +185,7 @@ type Executor struct {
 	squash                         bool
 	labels                         []string
 	annotations                    []string
+	onbuild                        []string
```
Comment (TomSweeneyRedHat) [inline imagebuildah/build.go]: I commented this out too, along with the call below to fill it up and it doesn't seem to be needed.

File: imagebuildah/build.go [hunk 5]
```diff
@@ -527,6 +530,7 @@ func NewExecutor(store storage.Store, options BuildOptions) (*Executor, error) {
 		squash:                options.Squash,
 		labels:                append([]string{}, options.Labels...),
 		annotations:           append([]string{}, options.Annotations...),
+		onbuild:               append([]string{}, options.OnBuild...),
```
Comment (TomSweeneyRedHat) [inline imagebuildah/build.go]: I was able to comment this out and it seemed to work without it.

Comment (rhatdan) [reply imagebuildah/build.go]: These fields are necessary to get 

buildah config --onbuild "RUN echo foobar" $CTR

To work.

**Comment by rhatdan** [comment]
@TomSweeneyRedHat I added a test to make sure `buildah config --onbuild ` works.

File: tests/bud.bats [hunk 6]
```diff
@@ -693,3 +693,49 @@ load helpers
...
+  run ls ${root}/onbuild1 ${root}/onbuild2
+  echo "$output"
+  [ "$status" -eq 0 ]
+  run ls ${root}/onbuild3
+  echo "$output"
+  [ "$status" -ne 0 ]
+  buildah umount ${cid}
+
+  target=onbuild-image3
+  run buildah --debug=false config --onbuild "RUN touch /onbuild4" ${cid}
```
Comment (TomSweeneyRedHat) [inline tests/bud.bats]: Think you need 
--signature-policy ${TESTSDIR}/policy.json 
here too.

**Comment by TomSweeneyRedHat** [comment]
@rhatdan I tested again and this time with commit and everything seemed to still work with those lines that I noted commented out.  I'm not seeing any effect of them running, I believe they're all no-ops.

Also in my review @nalind noted a concern that `buildah from` might not being handling "ONBUILD" correctly, and his concern seems founded.  The OnBuild command doesn't seem to be getting fired via that command, but if I build a Dockerfile with only "FROM onbuildimage" in it and then call `buildah bud` then the ONBUILD command gets fired.

File: tests/bud.bats (hunk 6 above)
Comment (rhatdan) [reply tests/bud.bats]: I think you only need signature-policy when doing a push or a pull.

File: imagebuildah/build.go (hunk 3 above)
Comment (rhatdan) [reply imagebuildah/build.go]: Removed

File: imagebuildah/build.go (hunk 4 above)
Comment (rhatdan) [reply imagebuildah/build.go]: Removed

**Comment by rhatdan** [comment]
@TomSweeneyRedHat If I remove ClearOnBuild, I see the second container image that I base on the first container image include the first images onbuild lines.

File: tests/bud.bats (hunk 6 above)
Comment (TomSweeneyRedHat) [reply tests/bud.bats]: My bad, yep, not needed for config and it looks to be blowing up the tests.  You do need it for the commit a few lines down,  Sorry for the dyslexic moment.    bud, commit, from and push commands require it in the buds tests.

File: tests/bud.bats [hunk 7]
```diff
@@ -693,3 +693,49 @@ load helpers
...
+  echo "$output"
+  [ "$status" -ne 0 ]
+  buildah umount ${cid}
+
+  target=onbuild-image3
+  run buildah --debug=false config --signature-policy ${TESTSDIR}/policy.json --onbuild "RUN touch /onbuild4" ${cid}
+  echo "$output"
+  [ "$status" -eq 0 ]
+
+  buildah commit --format docker ${cid} ${target}
```
Comment (TomSweeneyRedHat) [inline tests/bud.bats]: This however is going to need --signature-policy.

Comment (rhatdan) [reply tests/bud.bats]: yup

File: tests/bud.bats [hunk 8]
```diff
@@ -693,3 +693,49 @@ load helpers
...
+  [ "$status" -eq 0 ]
+  [ "$output" = '["RUN touch /onbuild3"]' ]
+  cid=$(buildah from ${target})
+  root=$(buildah mount ${cid})
+  run ls ${root}/onbuild1 ${root}/onbuild2
+  echo "$output"
+  [ "$status" -eq 0 ]
+  run ls ${root}/onbuild3
+  echo "$output"
+  [ "$status" -ne 0 ]
```
Comment (TomSweeneyRedHat) [inline tests/bud.bats]: Redhat CI is staying that line 725 failed.

Comment (TomSweeneyRedHat) [reply tests/bud.bats]: running the test by hand, I think line 725 s/b '-eq 0' as the directory is there.

File: tests/bud.bats [hunk 9]
```diff
@@ -693,3 +693,49 @@ load helpers
...
+  target=onbuild-image3
+  run buildah --debug=false config --onbuild "RUN touch /onbuild4" ${cid}
+  echo "$output"
+  [ "$status" -eq 0 ]
+
+  buildah commit --signature-policy ${TESTSDIR}/policy.json --format docker ${cid} ${target}
+  run buildah --debug=false inspect --format '{{printf "%q" .Docker.Config.OnBuild}}' ${target}
+  echo "$output"
+  [ "$status" -eq 0 ]
+  [ "$output" = '["RUN touch /onbuild3" "RUN touch /onbuild4"]' ]
```
Comment (TomSweeneyRedHat) [inline tests/bud.bats]: Should the output be just "RUN touch /onbuidl4" here?  I thought that's what the clearonbuild function did.

File: tests/bud.bats (hunk 8 above)
Comment (rhatdan) [reply tests/bud.bats]: Yes the buildah from changes are causing the tests to blow up.  Fixing up tests.

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please.

**Comment by rhatdan** [comment]
@umohnani8 @TomSweeneyRedHat @mheon @ipbabble PTAL
Finally passes tests.

**Comment by rhatdan** [comment]
@giuseppe @mheon @umohnani8 PTAL

File: cmd/buildah/from.go [hunk 10]
```diff
@@ -67,6 +68,97 @@ var (
...
+			builder.SetUser(strings.Join(args, " "))
+			continue
+		case "VOLUME":
+			builder.AddVolume(strings.Join(args, " "))
+			continue
+		case "WORKINGDIR":
+			builder.SetWorkDir(strings.Join(args, " "))
+			continue
+		default:
+			logrus.Errorf("OnBuild error %q invalid", onBuildSpec)
```
Comment (mheon) [inline cmd/buildah/from.go]: This error seems really poor - can you reword it to something like `OnBuild command %q is invalid`?

**Comment by mheon** [comment]
LGTM minus wording on one error message

**Comment by rhatdan** [comment]
@rh-atomic-bot r=mheon

**Comment by umohnani8** [comment]
LGTM
